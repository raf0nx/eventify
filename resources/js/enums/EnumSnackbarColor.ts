export enum EnumSnackbarColor {
    SUCCESS = "deep-purple",
    INFO = "blue",
    WARNING = "orange darken-1",
    ERROR = "red accent-2"
}
