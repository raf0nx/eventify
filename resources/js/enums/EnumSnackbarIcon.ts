export enum EnumSnackbarIcon {
    SUCCESS = "mdi-check-circle",
    INFO = "mdi-information-outline",
    WARNING = "mdi-alert",
    ERROR = "mdi-alert-circle-outline"
}
