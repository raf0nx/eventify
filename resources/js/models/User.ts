export interface User {
    id: Number;
    name: String;
    email: string;
    created_at: Date;
    email_verified_at?: Date;
    updated_at: Date;
}
