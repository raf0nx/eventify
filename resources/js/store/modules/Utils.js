export const namespaced = true;

export const state = {
    showSnackbar: false,
    showAlert: false,
    loading: false
};
export const mutations = {
    SET_ALERT(state, setAlert) {
        state.showAlert = setAlert;
    },
    SET_LOADING(state, isLoading) {
        state.loading = isLoading;
    },
    SET_SNACKBAR(state, setSnackbar) {
        state.showSnackbar = setSnackbar;
    }
};
export const actions = {
    setSnackbar({ commit }) {
        commit("SET_SNACKBAR", true);
        setTimeout(() => commit("SET_SNACKBAR", false), 2500);
    }
};

export const getters = {
    showAlert: state => {
        return state.showAlert;
    },
    isLoading: state => {
        return state.loading;
    },
    showSnackbar: state => {
        return state.showSnackbar;
    }
};
