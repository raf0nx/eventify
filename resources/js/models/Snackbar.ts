import { EnumSnackbarColor } from "@/enums/EnumSnackbarColor";
import { EnumSnackbarIcon } from "@/enums/EnumSnackbarIcon";

export class SnackbarModel {
    showSnackbar = false;
    message = "";
    icon = EnumSnackbarIcon.SUCCESS;
    color = EnumSnackbarColor.SUCCESS;

    setShowSnackbar(value: boolean): SnackbarModel {
        this.showSnackbar = value;
        return this;
    }

    setMessage(message: string): SnackbarModel {
        this.message = message;
        return this;
    }

    setIcon(icon: EnumSnackbarIcon): SnackbarModel {
        this.icon = icon;
        return this;
    }

    setColor(color: EnumSnackbarColor): SnackbarModel {
        this.color = color;
        return this;
    }
}
