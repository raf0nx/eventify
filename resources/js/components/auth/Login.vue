<template>
	<v-col
		cols="12"
		md="5"
		class="d-flex flex-column justify-center px-12"
		dark
	>
		<v-card elevation="0">
			<h1
				class="
					main-title
					font-weight-bold
					deep-purple--text
					text--darken-4
				"
			>
				Sign In
			</h1>
			<ValidationObserver ref="form" v-slot="{ passes, invalid }">
				<v-form class="py-12" @submit.prevent="passes(loginUser)">
					<ValidationProvider
						v-slot="{ errors }"
						rules="required|email"
						name="E-mail"
					>
						<v-text-field
							color="deep-purple"
							type="email"
							label="Email Address"
							data-cy="email"
							:error-messages="errors"
							v-model="formData.email"
						></v-text-field>
					</ValidationProvider>
					<ValidationProvider
						v-slot="{ errors }"
						rules="required"
						name="Password"
						vid="password"
					>
						<v-text-field
							color="deep-purple"
							type="password"
							label="Password"
							data-cy="password"
							:error-messages="errors"
							v-model="formData.password"
						></v-text-field>
					</ValidationProvider>
					<v-btn
						color="deep-purple"
						type="submit"
						class="white--text font-weight-bold mt-12"
						data-cy="submit"
						large
						block
						:disabled="invalid"
						>Continue</v-btn
					>
				</v-form>
			</ValidationObserver>
			<p class="text-center">
				Don't have an account?
				<v-btn :to="{ name: 'Register' }" plain text small>
					Sign up
				</v-btn>
			</p>
		</v-card>
	</v-col>
</template>

<script lang="ts">
	import { Vue, Component } from "vue-property-decorator";
	import { ValidationProvider, ValidationObserver } from "vee-validate";

	import AuthService from "@/services/AuthService";
	import { AuthModule } from "@modules/Auth";
	import { UtilsModule } from "@modules/Utils";
	import { LoginFormData } from "./types";

	@Component({
		components: { ValidationProvider, ValidationObserver },
	})
	export default class Login extends Vue {
		formData: LoginFormData = { email: "", password: "" };
		$refs!: {
			form: InstanceType<typeof ValidationObserver>;
		};

		async loginUser(): Promise<void> {
			UtilsModule.setLoading(true);
			try {
				await AuthService.loginUser(this.formData);
				const authUser = await AuthModule.getAuthUser();
				if (authUser) {
					this.$router.push({ name: "Home" });
				} else {
					throw new Error("Cannot authenticate user");
				}
			} catch (error) {
				const errorsData = error.response.data.errors;
				this.$refs.form.setErrors({
					password: errorsData.password,
				});
			}
			UtilsModule.setLoading(false);
		}
	}
</script>

<style scoped>
</style>