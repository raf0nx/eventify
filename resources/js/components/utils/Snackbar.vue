<template>
	<v-snackbar
		v-model="snackbar.showSnackbar"
        class="mb-8"
		app
		fixed
		bottom
		:color="snackbar.color"
	>
		<div class="d-flex align-center">
			<v-icon color="white">{{ snackbar.icon }}</v-icon>
			<span class="font-weight-bold ml-2">{{ snackbar.message }}</span>
		</div>
	</v-snackbar>
</template>

<script lang="ts">
	import { Vue, Component } from "vue-property-decorator";

	import { UtilsModule } from "@modules/Utils";
	import { SnackbarModel } from "@/models/Snackbar";

	@Component
	export default class Snackbar extends Vue {
		get snackbar(): SnackbarModel {
			return UtilsModule.snackbar;
		}
	}
</script>

<style scoped>
</style>