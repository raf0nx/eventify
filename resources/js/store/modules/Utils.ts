import {
    Module,
    VuexModule,
    Mutation,
    Action,
    getModule
} from "vuex-module-decorators";

import store from "@/store/store";
import { SnackbarModel } from "@/models/Snackbar";
import { AlertModel } from "@/models/Alert";

export interface UtilsState {
    snackbarNotification: SnackbarModel;
    alertNotification: AlertModel;
    isLoading: boolean;
    navDrawer: boolean;
}

@Module({ dynamic: true, store, name: "utils" })
class Utils extends VuexModule implements UtilsState {
    snackbarNotification = new SnackbarModel();
    alertNotification = new AlertModel();
    isLoading = false;
    navDrawer = false;

    @Mutation
    private SET_ALERT(alert: AlertModel): void {
        this.alertNotification = alert;
    }

    @Mutation
    private SET_LOADER(isLoading: boolean): void {
        this.isLoading = isLoading;
    }

    @Mutation
    private SET_SNACKBAR(snackbar: SnackbarModel): void {
        this.snackbarNotification = snackbar;
    }

    @Mutation
    private SET_NAV_DRAWER(setNavDrawer: boolean): void {
        this.navDrawer = setNavDrawer;
    }

    @Action
    setAlert(alert: AlertModel): void {
        this.SET_ALERT(alert);
    }

    @Action
    setLoader(isLoading: boolean): void {
        this.SET_LOADER(isLoading);
    }

    @Action
    setSnackbar(snackbar: SnackbarModel): void {
        this.SET_SNACKBAR(snackbar);
    }

    @Action
    setNavDrawer(setNavDrawer: boolean): void {
        this.SET_NAV_DRAWER(setNavDrawer);
    }

    get alert(): AlertModel {
        return this.alertNotification;
    }

    get loader(): boolean {
        return this.isLoading;
    }

    get snackbar(): SnackbarModel {
        return this.snackbarNotification;
    }

    get isNavDrawer(): boolean {
        return this.navDrawer;
    }
}

export const UtilsModule = getModule(Utils);
