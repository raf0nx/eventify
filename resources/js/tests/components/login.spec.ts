import { createLocalVue, mount, Wrapper } from "@vue/test-utils";
import Vuetify from "vuetify";
import axios from "axios";
import VueRouter from "vue-router";

import Login from "@components/auth/Login.vue";
import { user } from "@/tests/constans/User";
import { UtilsModule } from "@/store/modules/Utils";
import { EnumSnackbarColor } from "@/enums/EnumSnackbarColor";
import { EnumSnackbarIcon } from "@/enums/EnumSnackbarIcon";
import { SnackbarModel } from "@/models/Snackbar";

jest.mock("axios");

describe("Login.vue", () => {
    const localVue = createLocalVue();
    localVue.use(VueRouter);
    const router = new VueRouter();
    let wrapper: Wrapper<Login>;
    let vuetify: Vuetify;

    beforeEach(() => {
        vuetify = new Vuetify();
        wrapper = mount(Login, {
            localVue,
            vuetify,
            router
        });
    });

    afterEach(() => {
        jest.clearAllMocks();
        wrapper.destroy();
    });

    it("Should match snapshot", () => {
        // Assert
        expect(wrapper.html()).toMatchSnapshot();
    });

    it("Should log in user successfully", async () => {
        // Arrange
        // @ts-ignore
        axios.post.mockImplementationOnce(() => Promise.resolve());
        // @ts-ignore
        axios.get.mockImplementationOnce(() => Promise.resolve({ data: user }));

        // Act
        // @ts-ignore
        await wrapper.vm.loginUser();

        // Assert
        // @ts-ignore
        expect(wrapper.vm.$router.currentRoute.path === "/");
    });

    it("Should fail retrieving user after logging in", async () => {
        // Arrange
        // @ts-ignore
        axios.post.mockImplementationOnce(() => Promise.resolve());
        // @ts-ignore
        axios.get.mockImplementationOnce(() => Promise.resolve({ data: null }));

        // Act
        // @ts-ignore
        await wrapper.vm.loginUser();

        // Assert
        // @ts-ignore
        expect(UtilsModule.snackbar).toEqual(
            new SnackbarModel()
                .setShowSnackbar(true)
                .setMessage("Cannot authenticate you. Please try again.")
                .setColor(EnumSnackbarColor.ERROR)
                .setIcon(EnumSnackbarIcon.ERROR)
        );
    });

    it("Should fail loggin in user", async () => {
        // Arrange
        const error = {
            response: {
                data: {
                    errors: {
                        password: ["Email or password are incorrect!"]
                    }
                }
            }
        };

        // @ts-ignore
        axios.post.mockImplementationOnce(() => Promise.reject(error));

        // Act
        // @ts-ignore
        await wrapper.vm.loginUser();

        // Assert
        // @ts-ignore
        expect(wrapper.vm.loginUser()).rejects;
    });
});
