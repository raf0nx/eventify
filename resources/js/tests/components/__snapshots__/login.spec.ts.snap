// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login.vue Should match snapshot 1`] = `
"<div dark=\\"\\" class=\\"d-flex flex-column justify-center px-12 col-md-5 col-12\\">
  <div class=\\"v-card v-sheet theme--light elevation-0\\">
    <h1 class=\\"
				auth-form__title
				font-weight-bold
				deep-purple--text
				text--darken-4
			\\">
      Sign In
    </h1> <span><form novalidate=\\"novalidate\\" class=\\"v-form py-12\\"><span><div class=\\"v-input theme--light v-text-field\\"><div class=\\"v-input__control\\"><div class=\\"v-input__slot\\"><div class=\\"v-text-field__slot\\"><label for=\\"input-11\\" class=\\"v-label theme--light\\" style=\\"left: 0px; position: absolute;\\">Email Address</label><input data-cy=\\"email\\" id=\\"input-11\\" type=\\"email\\"></div></div><div class=\\"v-text-field__details\\"><div class=\\"v-messages theme--light\\"><transition-group-stub tag=\\"div\\" name=\\"message-transition\\" class=\\"v-messages__wrapper\\"></transition-group-stub></div></div></div></div></span> <span><div class=\\"v-input theme--light v-text-field\\"><div class=\\"v-input__control\\"><div class=\\"v-input__slot\\"><div class=\\"v-text-field__slot\\"><label for=\\"input-15\\" class=\\"v-label theme--light\\" style=\\"left: 0px; position: absolute;\\">Password</label><input data-cy=\\"password\\" id=\\"input-15\\" type=\\"password\\"></div></div><div class=\\"v-text-field__details\\"><div class=\\"v-messages theme--light\\"><transition-group-stub tag=\\"div\\" name=\\"message-transition\\" class=\\"v-messages__wrapper\\"></transition-group-stub></div></div></div></div></span> <button type=\\"submit\\" class=\\"white--text font-weight-bold mt-12 v-btn v-btn--block v-btn--is-elevated v-btn--has-bg theme--light v-size--large deep-purple\\" data-cy=\\"submit\\"><span class=\\"v-btn__content\\">Continue</span></button></form></span>
    <p class=\\"text-center\\">
      Don't have an account?
      <a href=\\"#/\\" aria-current=\\"page\\" class=\\"v-btn--active v-btn v-btn--plain v-btn--router v-btn--text theme--light v-size--small\\"><span class=\\"v-btn__content\\">
				Sign up
			</span></a></p>
  </div>
</div>"
`;
